### ISSUE TITLE
Issue with jQuery draggable

### ISSUE THREAD

**Issue by flynhigher** [angular-ui/ui-sortable#354]
Hi,

First of all, thanks for creating this wonderful directive.
I'm using this directive a lot and it's invaluable tool for web ui development in Angular.

Recently, I received a requirement to drag an item from a listbox, which should not have sorting feature, into the sortable control.
So I created "draggable" directive and provided event handler and handle the "update" event in sortable by adding a dragged item in the list to which ui-sortable's ng-model is referring.
Everything worked fine and then weird thing happens: when only two items are added by dragging sortable doesn't work anymore and as soon as you add one more item it works again.

The way I implemented is: "draggable" is adding list item by modifying dom inside ng-repeat when dragged. So I had to remove that in "update" event and then insert into the collection. I'm thinking that ng-repeat can't handle correctly when dom modification happens without its knowledge but other than that I don't have any clue why.

Did anyone experience like this before or try to use ui-sortable with draggable?
If you can think of any solution to this problem let me know.

Thanks,
Terry

**Comment by flynhigher** [comment]
FYI, I tried connected sortable but I want to disable sorting and copying instead of moving from source list, which is not working I believe.

**Comment by thgreasi** [comment]
The way I've recently done it, is similar to what is shown in the cloning
example.
The key difference is that I need to create one sortable per each element I
want to be draggable. If you try to solve it this way, then I would suggest
you to use different options objects for the sortable and the emulated
draggable elements.

It's not the most elegant solution but it works,  since the
jquery-ui-sortable component is a superset/combination of the draggable and
droppable component.
Some of the pros thought are:
* you will be able use the same extensive api for both parts of the
application
* you will not have to find or program a new module by yourself
* you will not have to do extensive testing
* you will not face interoperability issues

Moreover you should avoid premature optimization and initially make things
just work. If this is not good enough for you then you could investigating
writing an interoperable draggable component by yourself.

If you face any difficulties, I might be able to create an example codepen
for this use case.

**Comment by paulflo150** [comment]
Thank you as well for the excellent work.

Could you please create that codepen, I too am interested in this?

Thanks!

**Comment by thgreasi** [comment]
[Here](http://codepen.io/thgreasi/pen/LVVJgK) is an example of what I was talking about.
I'm going to also add it to README.md.

**Comment by paulflo150** [comment]
Great thanks! 

And it is not possible at this time to leave the item in place and/ or perhaps add it back when the drag starts correct?

**Comment by thgreasi** [comment]
I think that this is possible by using the helper option as a function.
Something like: returning a clone of the dragged element and setting the
original item as visible. Take a look at the jquery-ui api page for more
info.

**Comment by paulflo150** [comment]
Excellent, I got it working, could you take a look to make sure I am not doing anything I shouldn't be?

http://codepen.io/anon/pen/aOvpZm

I added the following:
helper: function(e, ui) {
      return ui.clone();
    },
    start: function(e, ui) {
      $(ui.item).show();      
    },

and in the css:
.draggable-element-container .ui-sortable-placeholder {
  display: none;
}

Thanks,
Paul

**Comment by thgreasi** [comment]
This looks ok.

Just as a code clarity suggestion, I think that the second parameter of the
helper callback is `item` and should read:
helper: function(e, item) {
return item.clone();
}

But of course the code has the same results and the same meaning for the
computer.

I will check it when I'm home.

**Comment by paulflo150** [comment]
Thank you sir.

One last problem I ran into. I have a nested ng repeat using the dynamic directive, and I can't quite figure out how to track by $index correctly (or perhaps I am using the directive the wrong way - I can create a plunker unless it may be something obvious). 

This is what my templates look like:

        <div data-ng-repeat="item in rootItem.items">
            <tg-dynamic-directive ng-model="item" tg-dynamic-directive-view="getView(rootItem)">
            </tg-dynamic-directive>
        </div>

        <script type="text/ng-template" id="bootstrap.html">
            <div class="connected-drop-target-sortable" data-ui-sortable="sortableOptions" ng-model="ngModelItem" style="min-height: 100px;">
                <div ng-repeat="item in ngModelItem">
                    <div data-ng-if="item.view === 'bootstrap.html'">
                        <div data-ng-repeat="index in item.items" class="column" data-ng-class="'col-md-' + 12/item.items.length">
                            <tg-dynamic-directive ng-model="index" tg-dynamic-directive-view="getView(item)">
                            </tg-dynamic-directive>
                        </div>
                    </div>
                    <div data-ng-if="item.view === 'text.html'">
                        <tg-dynamic-directive ng-model="item" tg-dynamic-directive-view="getView(item)">
                        </tg-dynamic-directive>
                    </div>
                </div>
            </div>
        </script>
        <script type="text/ng-template" id="text.html">
            <div style="display: block;">
                <h4>This is a text</h4>
            </div>
        </script>

    $scope.rootItem = {
        title: 'Root Item',
        view: 'bootstrap.html',
        items:
            [[{
                title: 'Bootstrap one column',
                view: 'bootstrap.html',
                items: [[]]
            },
            {
                title: 'Bootstrap two columns',
                view: 'bootstrap.html',
                items: [[], []]
            }
            ,
            {
                title: 'Title',
                view: 'title.html'
            }
            ]]
    };

Then I have a toolbox that I can drag/ drop items from onto my page - if it's a bootstrap control it will generate the desired number of columns which will either accept another bootstrap template or the title. The way it is now if I drop more than one into the same container angular will complain about duplicates, so I thought I would create a copy of the model when I push it back into the toolbox. This seems to work most of the time, although on occasion sections of items will dissapear when dragging other sections (I wish I could make it happen on demand, but this only happens rarely). So I thought that may be because angular is confused and tried adding a track by. If I add a track by $index however to both ng repeats in bootstrap.html, all my items seem to be bound to the same scope, they pretty much all look the same

Any thoughts on what I may be doing wrong?

Thanks,

**Comment by thgreasi** [comment]
An example would make it easier for me to locate, fix and present the
changes needed.

Generally, (deep) cloning a dropped object that is not moved. That's
reasonable especially in the case that we try to emulate the
draggable-sortable interaction, since otherwise the result would be having
several references to the same object.

Generally I would suggest to use getview on the item of each repetition
instead of the iterated collection.
In other words, I would expect something like:
<div data-ng-repeat="item in rootItem.items track by $index">
<tg-dynamic-directive ng-model="item"
tg-dynamic-directive-view="getView(item)"> </tg-dynamic-directive> </div>

**Comment by paulflo150** [comment]
Yup, should have done that to begin with :), here is the pen: http://codepen.io/anon/pen/aOvJjR?editors=101

I would certainly like to be able to follow your suggestion and change it to the way your dynamic example is set up, and I ended up with this because I was having a little trouble splitting up the columns. It would also be nice to be able to get rid of those ng-ifs and let getView return the view it needs to return. I will take another shot at it as well, and perhaps you could help me out if you get some free time.

Thanks,
Paul

**Comment by thgreasi** [comment]
Could not isolate any error without the `$track by`, will try with it.

**Comment by thgreasi** [comment]
There is obviously something wrong with track by `$index`...
Issues with it have already been reported in other issues of this repo.

I tend to use an ID for each of my objects since they are representations of backend objects.
I would suggest you to do the same, I like to use [node-uuid](https://github.com/broofa/node-uuid).

If in the meantime how find out what's the case, I would be more than happy if you could report it back in here.

**Comment by flynhigher** [comment]
Fantastic, it works!
Thank you, Thodoris and Paul!

One modification that I had to do is that I added one more layer between original collection to nested ui-sortable collection in order to accommodate filters.
```html
<div ng-repeat="x in originalDraggables" ng-init="draggableArray = [x]">
    <div ui-sortable="draggableOptions" class="screen draggable-element-container"
        ng-model="draggableArray">
          <div ng-repeat="draggable in draggableArray" class="app">{{draggable.title}}</div>
    </div>
</div>
```

**Comment by paulflo150** [comment]
Got a little distracted with ... oh yeah work :), but I should be getting back to my side project shortly and I will let you know if I figure out the index problem, although if I run into issues I will follow your suggestion first and look at uuid. So far just copying everything back into the toolbox seems to work very well though. There is one exception, but I do not know how to reproduce it which happens from time to time and it adds a null object to the collection - I have made some changes since I had seen it last, so I am hopping I was just doing something wrong before :)

When generating an ID yourself, do you also explicitly tell it to track by that id, or would that be unnecessary? (i.e.: data-ng-repeat="item in items track by item.myuuid" or just leave it as data-ng-repeat="item in items" and angular should be ok since the objects will be clearly different?)

Thanks again for all your help, it is very refreshing to actually see a project owner go to great lengths to help out!

Paul

**Comment by thgreasi** [comment]
Happy you are back.
This it's not absolutely true/correct but: Always `track by` if you can!!
(and doing so doesn't break your implementation.)

`track by` originally caused some issues & confusion to lots of devs and
repos (since it changed the original semantics of the ngRepeat), but it's
generally the way to go since it will try to perform the minimum DOM
operations when a collection changes.

Since angular v1.2 was released all the tests and examples of this repo
also use it.
